File: cc-native-compiler-osaka-202605/src/typescript/main.ts
// nb-lang compiler — TypeScript reference implementation (Phase 1)
// 実行: bun run main.ts <source.nb>

import * as fs from "node:fs";

// ===== AST =====

type Expr =
  | { kind: "IntLit"; value: number }
  | { kind: "Var"; name: string }
  | { kind: "BinOp"; op: string; left: Expr; right: Expr };

type Stmt =
  | { kind: "ValDef"; name: string; init: Expr }
  | { kind: "VarDef"; name: string; init: Expr }
  | { kind: "Reassign"; name: string; value: Expr }
  | { kind: "If"; cond: Expr; thenBlk: Stmt[]; elseBlk: Stmt[] }
  | { kind: "While"; cond: Expr; body: Stmt[] }
  | { kind: "Print"; value: Expr };

// ===== Lexer =====

type Token =
  | { kind: "Int"; value: number }
  | { kind: "Ident"; value: string }
  | { kind: "Kw"; value: "val" | "var" | "if" | "else" | "while" | "print" }
  | { kind: "Op"; value: string }
  | { kind: "LParen" }
  | { kind: "RParen" }
  | { kind: "LBrace" }
  | { kind: "RBrace" }
  | { kind: "Semi" }
  | { kind: "EOF" };

const KEYWORDS = new Set(["val", "var", "if", "else", "while", "print"]);

function tokenize(src: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  const n = src.length;

  while (i < n) {
    const c = src[i];
    if (/\s/.test(c)) { i++; continue; }
    if (c === "/" && src[i + 1] === "/") {
      while (i < n && src[i] !== "\n") i++;
      continue;
    }
    if (/\d/.test(c)) {
      let j = i;
      while (j < n && /\d/.test(src[j])) j++;
      tokens.push({ kind: "Int", value: parseInt(src.substring(i, j)) });
      i = j;
      continue;
    }
    if (/[a-zA-Z_]/.test(c)) {
      let j = i;
      while (j < n && /[a-zA-Z0-9_]/.test(src[j])) j++;
      const word = src.substring(i, j);
      if (KEYWORDS.has(word)) {
        tokens.push({ kind: "Kw", value: word as "val" | "var" | "if" | "else" | "while" | "print" });
      } else {
        tokens.push({ kind: "Ident", value: word });
      }
      i = j;
      continue;
    }
    if (c === "(") { tokens.push({ kind: "LParen" }); i++; continue; }
    if (c === ")") { tokens.push({ kind: "RParen" }); i++; continue; }
    if (c === "{") { tokens.push({ kind: "LBrace" }); i++; continue; }
    if (c === "}") { tokens.push({ kind: "RBrace" }); i++; continue; }
    if (c === ";") { tokens.push({ kind: "Semi" }); i++; continue; }
    if (c === "=" && src[i + 1] === "=") { tokens.push({ kind: "Op", value: "==" }); i += 2; continue; }
    if (c === "!" && src[i + 1] === "=") { tokens.push({ kind: "Op", value: "!=" }); i += 2; continue; }
    if (c === "<" && src[i + 1] === "=") { tokens.push({ kind: "Op", value: "<=" }); i += 2; continue; }
    if (c === ">" && src[i + 1] === "=") { tokens.push({ kind: "Op", value: ">=" }); i += 2; continue; }
    if ("+-*/%<>=".includes(c)) { tokens.push({ kind: "Op", value: c }); i++; continue; }
    throw new Error(`Unknown char: '${c}' at position ${i}`);
  }
  tokens.push({ kind: "EOF" });
  return tokens;
}

// ===== Parser =====

class Parser {
  private pos = 0;

  constructor(private readonly tokens: Token[]) {}

  private peek(): Token { return this.tokens[this.pos]; }
  private consume(): Token { return this.tokens[this.pos++]; }

  parseProgram(): Stmt[] {
    const stmts: Stmt[] = [];
    while (this.peek().kind !== "EOF") {
      stmts.push(this.parseStmt());
    }
    return stmts;
  }

  private parseStmt(): Stmt {
    const t = this.peek();
    if (t.kind === "Kw") {
      switch (t.value) {
        case "val":   return this.parseValDef();
        case "var":   return this.parseVarDef();
        case "if":    return this.parseIf();
        case "while": return this.parseWhile();
        case "print": return this.parsePrint();
      }
    }
    if (t.kind === "Ident") return this.parseReassign();
    throw new Error(`Unexpected token: ${JSON.stringify(t)}`);
  }

  private parseValDef(): Stmt {
    this.consume(); // val
    const name = this.consumeIdent();
    this.expectOp("=");
    const init = this.parseExpr();
    this.expectSemi();
    return { kind: "ValDef", name, init };
  }

  private parseVarDef(): Stmt {
    this.consume(); // var
    const name = this.consumeIdent();
    this.expectOp("=");
    const init = this.parseExpr();
    this.expectSemi();
    return { kind: "VarDef", name, init };
  }

  private parseReassign(): Stmt {
    const name = this.consumeIdent();
    this.expectOp("=");
    const value = this.parseExpr();
    this.expectSemi();
    return { kind: "Reassign", name, value };
  }

  private parseIf(): Stmt {
    this.consume(); // if
    this.expectKind("LParen");
    const cond = this.parseExpr();
    this.expectKind("RParen");
    const thenBlk = this.parseBlock();
    let elseBlk: Stmt[] = [];
    const next = this.peek();
    if (next.kind === "Kw" && next.value === "else") {
      this.consume();
      elseBlk = this.parseBlock();
    }
    return { kind: "If", cond, thenBlk, elseBlk };
  }

  private parseWhile(): Stmt {
    this.consume(); // while
    this.expectKind("LParen");
    const cond = this.parseExpr();
    this.expectKind("RParen");
    const body = this.parseBlock();
    return { kind: "While", cond, body };
  }

  private parsePrint(): Stmt {
    this.consume(); // print
    this.expectKind("LParen");
    const value = this.parseExpr();
    this.expectKind("RParen");
    this.expectSemi();
    return { kind: "Print", value };
  }

  private parseBlock(): Stmt[] {
    this.expectKind("LBrace");
    const stmts: Stmt[] = [];
    while (this.peek().kind !== "RBrace") {
      stmts.push(this.parseStmt());
    }
    this.consume(); // }
    return stmts;
  }

  private parseExpr(): Expr { return this.parseComparison(); }

  private parseComparison(): Expr {
    let left = this.parseAdditive();
    while (this.isOp(["==", "!=", "<", "<=", ">", ">="])) {
      const op = (this.consume() as { kind: "Op"; value: string }).value;
      const right = this.parseAdditive();
      left = { kind: "BinOp", op, left, right };
    }
    return left;
  }

  private parseAdditive(): Expr {
    let left = this.parseMultiplicative();
    while (this.isOp(["+", "-"])) {
      const op = (this.consume() as { kind: "Op"; value: string }).value;
      const right = this.parseMultiplicative();
      left = { kind: "BinOp", op, left, right };
    }
    return left;
  }

  private parseMultiplicative(): Expr {
    let left = this.parsePrimary();
    while (this.isOp(["*", "/", "%"])) {
      const op = (this.consume() as { kind: "Op"; value: string }).value;
      const right = this.parsePrimary();
      left = { kind: "BinOp", op, left, right };
    }
    return left;
  }

  private parsePrimary(): Expr {
    const t = this.consume();
    if (t.kind === "Int") return { kind: "IntLit", value: t.value };
    if (t.kind === "Ident") return { kind: "Var", name: t.value };
    if (t.kind === "LParen") {
      const e = this.parseExpr();
      this.expectKind("RParen");
      return e;
    }
    throw new Error(`Expected primary, got ${JSON.stringify(t)}`);
  }

  private isOp(values: string[]): boolean {
    const t = this.peek();
    return t.kind === "Op" && values.includes(t.value);
  }

  private consumeIdent(): string {
    const t = this.consume();
    if (t.kind !== "Ident") throw new Error(`Expected identifier, got ${JSON.stringify(t)}`);
    return t.value;
  }

  private expectOp(value: string): void {
    const t = this.consume();
    if (t.kind !== "Op" || t.value !== value) {
      throw new Error(`Expected '${value}', got ${JSON.stringify(t)}`);
    }
  }

  private expectKind(kind: Token["kind"]): void {
    const t = this.consume();
    if (t.kind !== kind) throw new Error(`Expected ${kind}, got ${JSON.stringify(t)}`);
  }

  private expectSemi(): void { this.expectKind("Semi"); }
}

// ===== CodeGen =====

class CodeGen {
  private lines: string[] = [];
  private tempCounter = 0;
  private labelCounter = 0;

  private fresh(): string {
    this.tempCounter++;
    return `%t${this.tempCounter}`;
  }

  private freshLabel(prefix: string): string {
    this.labelCounter++;
    return `${prefix}${this.labelCounter}`;
  }

  private emit(line: string): void { this.lines.push(line); }

  private detectTriple(): string {
    return process.platform === "darwin"
      ? "arm64-apple-macosx15.0.0"
      : "x86_64-pc-linux-gnu";
  }

  gen(program: Stmt[]): string {
    this.emit("; ModuleID = 'nb-lang'");
    this.emit(`target triple = "${this.detectTriple()}"`);
    this.emit("");
    this.emit('@fmt = private constant [5 x i8] c"%ld\\0A\\00"');
    this.emit("declare i32 @printf(ptr, ...)");
    this.emit("");
    this.emit("define i32 @main() {");
    this.emit("entry:");
    for (const s of program) this.genStmt(s);
    this.emit("  ret i32 0");
    this.emit("}");
    return this.lines.join("\n") + "\n";
  }

  private genStmt(s: Stmt): void {
    switch (s.kind) {
      case "ValDef":
      case "VarDef": {
        this.emit(`  %${s.name} = alloca i64, align 8`);
        const v = this.genExpr(s.init);
        this.emit(`  store i64 ${v}, ptr %${s.name}`);
        break;
      }
      case "Reassign": {
        const v = this.genExpr(s.value);
        this.emit(`  store i64 ${v}, ptr %${s.name}`);
        break;
      }
      case "Print": {
        const v = this.genExpr(s.value);
        this.emit(`  call i32 (ptr, ...) @printf(ptr @fmt, i64 ${v})`);
        break;
      }
      case "If": {
        const c = this.genExpr(s.cond);
        const ci1 = this.fresh();
        this.emit(`  ${ci1} = icmp ne i64 ${c}, 0`);
        const thenL = this.freshLabel("then");
        const elseL = this.freshLabel("else");
        const endL  = this.freshLabel("ifend");
        this.emit(`  br i1 ${ci1}, label %${thenL}, label %${elseL}`);
        this.emit(`${thenL}:`);
        for (const st of s.thenBlk) this.genStmt(st);
        this.emit(`  br label %${endL}`);
        this.emit(`${elseL}:`);
        for (const st of s.elseBlk) this.genStmt(st);
        this.emit(`  br label %${endL}`);
        this.emit(`${endL}:`);
        break;
      }
      case "While": {
        const condL = this.freshLabel("cond");
        const bodyL = this.freshLabel("body");
        const exitL = this.freshLabel("exit");
        this.emit(`  br label %${condL}`);
        this.emit(`${condL}:`);
        const c = this.genExpr(s.cond);
        const ci1 = this.fresh();
        this.emit(`  ${ci1} = icmp ne i64 ${c}, 0`);
        this.emit(`  br i1 ${ci1}, label %${bodyL}, label %${exitL}`);
        this.emit(`${bodyL}:`);
        for (const st of s.body) this.genStmt(st);
        this.emit(`  br label %${condL}`);
        this.emit(`${exitL}:`);
        break;
      }
    }
  }

  private genExpr(e: Expr): string {
    switch (e.kind) {
      case "IntLit": return e.value.toString();
      case "Var": {
        const r = this.fresh();
        this.emit(`  ${r} = load i64, ptr %${e.name}`);
        return r;
      }
      case "BinOp": {
        const lv = this.genExpr(e.left);
        const rv = this.genExpr(e.right);
        const res = this.fresh();
        const arith: Record<string, string> = {
          "+": "add", "-": "sub", "*": "mul", "/": "sdiv", "%": "srem"
        };
        const cmp: Record<string, string> = {
          "==": "eq", "!=": "ne", "<": "slt", "<=": "sle", ">": "sgt", ">=": "sge"
        };
        if (arith[e.op]) {
          this.emit(`  ${res} = ${arith[e.op]} i64 ${lv}, ${rv}`);
        } else if (cmp[e.op]) {
          const tmp = this.fresh();
          this.emit(`  ${tmp} = icmp ${cmp[e.op]} i64 ${lv}, ${rv}`);
          this.emit(`  ${res} = zext i1 ${tmp} to i64`);
        } else {
          throw new Error(`Unknown op: ${e.op}`);
        }
        return res;
      }
    }
  }
}

// ===== Main =====

const args = process.argv.slice(2);
if (args.length === 0) {
  console.error("usage: bun run main.ts <source.nb>");
  process.exit(1);
}

const src = fs.readFileSync(args[0], "utf-8");
const tokens = tokenize(src);
const parser = new Parser(tokens);
const program = parser.parseProgram();
const codegen = new CodeGen();
const ir = codegen.gen(program);
process.stdout.write(ir);

File: cc-native-compiler-osaka-202605/src/java/Main.java
// nb-lang compiler — Java reference implementation (Phase 1)
// 実行: java Main.java <source.nb>

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Main {

    // ===== AST =====

    sealed interface Expr permits IntLit, Var, BinOp {}
    record IntLit(long value)                            implements Expr {}
    record Var(String name)                              implements Expr {}
    record BinOp(String op, Expr left, Expr right)       implements Expr {}

    sealed interface Stmt permits ValDef, VarDef, Reassign, If, While, Print {}
    record ValDef(String name, Expr init)                implements Stmt {}
    record VarDef(String name, Expr init)                implements Stmt {}
    record Reassign(String name, Expr value)             implements Stmt {}
    record If(Expr cond, List<Stmt> thenBlk, List<Stmt> elseBlk) implements Stmt {}
    record While(Expr cond, List<Stmt> body)             implements Stmt {}
    record Print(Expr value)                             implements Stmt {}

    // ===== Token =====

    sealed interface Token permits IntT, Ident, Kw, Op, LParen, RParen, LBrace, RBrace, Semi, EOF {}
    record IntT(long value)        implements Token {}
    record Ident(String name)      implements Token {}
    record Kw(String name)         implements Token {}
    record Op(String value)        implements Token {}
    record LParen()                implements Token {}
    record RParen()                implements Token {}
    record LBrace()                implements Token {}
    record RBrace()                implements Token {}
    record Semi()                  implements Token {}
    record EOF()                   implements Token {}

    // ===== Lexer =====

    static final Set<String> KEYWORDS = Set.of("val", "var", "if", "else", "while", "print");

    static List<Token> tokenize(String src) {
        List<Token> tokens = new ArrayList<>();
        int i = 0, n = src.length();

        while (i < n) {
            char c = src.charAt(i);
            if (Character.isWhitespace(c)) { i++; continue; }
            if (c == '/' && i + 1 < n && src.charAt(i + 1) == '/') {
                while (i < n && src.charAt(i) != '\n') i++;
                continue;
            }
            if (Character.isDigit(c)) {
                int j = i;
                while (j < n && Character.isDigit(src.charAt(j))) j++;
                tokens.add(new IntT(Long.parseLong(src.substring(i, j))));
                i = j;
                continue;
            }
            if (Character.isLetter(c) || c == '_') {
                int j = i;
                while (j < n && (Character.isLetterOrDigit(src.charAt(j)) || src.charAt(j) == '_')) j++;
                String word = src.substring(i, j);
                tokens.add(KEYWORDS.contains(word) ? new Kw(word) : new Ident(word));
                i = j;
                continue;
            }
            if (c == '(') { tokens.add(new LParen()); i++; continue; }
            if (c == ')') { tokens.add(new RParen()); i++; continue; }
            if (c == '{') { tokens.add(new LBrace()); i++; continue; }
            if (c == '}') { tokens.add(new RBrace()); i++; continue; }
            if (c == ';') { tokens.add(new Semi());   i++; continue; }
            if (c == '=' && i + 1 < n && src.charAt(i + 1) == '=') { tokens.add(new Op("==")); i += 2; continue; }
            if (c == '!' && i + 1 < n && src.charAt(i + 1) == '=') { tokens.add(new Op("!=")); i += 2; continue; }
            if (c == '<' && i + 1 < n && src.charAt(i + 1) == '=') { tokens.add(new Op("<=")); i += 2; continue; }
            if (c == '>' && i + 1 < n && src.charAt(i + 1) == '=') { tokens.add(new Op(">=")); i += 2; continue; }
            if ("+-*/%<>=".indexOf(c) >= 0) { tokens.add(new Op(String.valueOf(c))); i++; continue; }
            throw new RuntimeException("Unknown char: '" + c + "' at " + i);
        }
        tokens.add(new EOF());
        return tokens;
    }

    // ===== Parser =====

    static class Parser {
        private final List<Token> tokens;
        private int pos = 0;

        Parser(List<Token> tokens) { this.tokens = tokens; }

        private Token peek() { return tokens.get(pos); }
        private Token consume() { return tokens.get(pos++); }

        List<Stmt> parseProgram() {
            List<Stmt> stmts = new ArrayList<>();
            while (!(peek() instanceof EOF)) stmts.add(parseStmt());
            return stmts;
        }

        private Stmt parseStmt() {
            Token t = peek();
            if (t instanceof Kw kw) {
                return switch (kw.name()) {
                    case "val"   -> parseValDef();
                    case "var"   -> parseVarDef();
                    case "if"    -> parseIf();
                    case "while" -> parseWhile();
                    case "print" -> parsePrint();
                    default      -> throw new RuntimeException("Unexpected keyword: " + kw.name());
                };
            }
            if (t instanceof Ident) return parseReassign();
            throw new RuntimeException("Unexpected token: " + t);
        }

        private Stmt parseValDef() {
            consume(); // val
            String name = consumeIdent();
            expectOp("=");
            Expr init = parseExpr();
            expectSemi();
            return new ValDef(name, init);
        }

        private Stmt parseVarDef() {
            consume(); // var
            String name = consumeIdent();
            expectOp("=");
            Expr init = parseExpr();
            expectSemi();
            return new VarDef(name, init);
        }

        private Stmt parseReassign() {
            String name = consumeIdent();
            expectOp("=");
            Expr value = parseExpr();
            expectSemi();
            return new Reassign(name, value);
        }

        private Stmt parseIf() {
            consume(); // if
            expect(LParen.class);
            Expr cond = parseExpr();
            expect(RParen.class);
            List<Stmt> thenBlk = parseBlock();
            List<Stmt> elseBlk = new ArrayList<>();
            if (peek() instanceof Kw kw && kw.name().equals("else")) {
                consume();
                elseBlk = parseBlock();
            }
            return new If(cond, thenBlk, elseBlk);
        }

        private Stmt parseWhile() {
            consume(); // while
            expect(LParen.class);
            Expr cond = parseExpr();
            expect(RParen.class);
            List<Stmt> body = parseBlock();
            return new While(cond, body);
        }

        private Stmt parsePrint() {
            consume(); // print
            expect(LParen.class);
            Expr v = parseExpr();
            expect(RParen.class);
            expectSemi();
            return new Print(v);
        }

        private List<Stmt> parseBlock() {
            expect(LBrace.class);
            List<Stmt> stmts = new ArrayList<>();
            while (!(peek() instanceof RBrace)) stmts.add(parseStmt());
            consume(); // }
            return stmts;
        }

        private Expr parseExpr() { return parseComparison(); }

        private Expr parseComparison() {
            Expr left = parseAdditive();
            while (isOp(Set.of("==", "!=", "<", "<=", ">", ">="))) {
                String op = ((Op) consume()).value();
                Expr right = parseAdditive();
                left = new BinOp(op, left, right);
            }
            return left;
        }

        private Expr parseAdditive() {
            Expr left = parseMultiplicative();
            while (isOp(Set.of("+", "-"))) {
                String op = ((Op) consume()).value();
                Expr right = parseMultiplicative();
                left = new BinOp(op, left, right);
            }
            return left;
        }

        private Expr parseMultiplicative() {
            Expr left = parsePrimary();
            while (isOp(Set.of("*", "/", "%"))) {
                String op = ((Op) consume()).value();
                Expr right = parsePrimary();
                left = new BinOp(op, left, right);
            }
            return left;
        }

        private Expr parsePrimary() {
            Token t = consume();
            if (t instanceof IntT it) return new IntLit(it.value());
            if (t instanceof Ident id) return new Var(id.name());
            if (t instanceof LParen) {
                Expr e = parseExpr();
                expect(RParen.class);
                return e;
            }
            throw new RuntimeException("Expected primary, got " + t);
        }

        private boolean isOp(Set<String> ops) {
            return peek() instanceof Op op && ops.contains(op.value());
        }

        private String consumeIdent() {
            Token t = consume();
            if (!(t instanceof Ident id)) throw new RuntimeException("Expected ident, got " + t);
            return id.name();
        }

        private void expectOp(String value) {
            Token t = consume();
            if (!(t instanceof Op op) || !op.value().equals(value)) {
                throw new RuntimeException("Expected '" + value + "', got " + t);
            }
        }

        private void expect(Class<? extends Token> kind) {
            Token t = consume();
            if (!kind.isInstance(t)) {
                throw new RuntimeException("Expected " + kind.getSimpleName() + ", got " + t);
            }
        }

        private void expectSemi() { expect(Semi.class); }
    }

    // ===== CodeGen =====

    static class CodeGen {
        private final StringBuilder sb = new StringBuilder();
        private int tempCounter = 0;
        private int labelCounter = 0;

        private String fresh() {
            tempCounter++;
            return "%t" + tempCounter;
        }

        private String freshLabel(String prefix) {
            labelCounter++;
            return prefix + labelCounter;
        }

        private void emit(String line) { sb.append(line).append('\n'); }

        private String detectTriple() {
            String os = System.getProperty("os.name").toLowerCase();
            return os.contains("mac") ? "arm64-apple-macosx15.0.0" : "x86_64-pc-linux-gnu";
        }

        String gen(List<Stmt> program) {
            emit("; ModuleID = 'nb-lang'");
            emit("target triple = \"" + detectTriple() + "\"");
            emit("");
            emit("@fmt = private constant [5 x i8] c\"%ld\\0A\\00\"");
            emit("declare i32 @printf(ptr, ...)");
            emit("");
            emit("define i32 @main() {");
            emit("entry:");
            for (Stmt s : program) genStmt(s);
            emit("  ret i32 0");
            emit("}");
            return sb.toString();
        }

        private void genStmt(Stmt s) {
            if (s instanceof ValDef d) {
                emit("  %" + d.name() + " = alloca i64, align 8");
                String v = genExpr(d.init());
                emit("  store i64 " + v + ", ptr %" + d.name());
            } else if (s instanceof VarDef d) {
                emit("  %" + d.name() + " = alloca i64, align 8");
                String v = genExpr(d.init());
                emit("  store i64 " + v + ", ptr %" + d.name());
            } else if (s instanceof Reassign r) {
                String v = genExpr(r.value());
                emit("  store i64 " + v + ", ptr %" + r.name());
            } else if (s instanceof Print p) {
                String v = genExpr(p.value());
                emit("  call i32 (ptr, ...) @printf(ptr @fmt, i64 " + v + ")");
            } else if (s instanceof If i) {
                String c = genExpr(i.cond());
                String ci1 = fresh();
                emit("  " + ci1 + " = icmp ne i64 " + c + ", 0");
                String thenL = freshLabel("then");
                String elseL = freshLabel("else");
                String endL  = freshLabel("ifend");
                emit("  br i1 " + ci1 + ", label %" + thenL + ", label %" + elseL);
                emit(thenL + ":");
                for (Stmt st : i.thenBlk()) genStmt(st);
                emit("  br label %" + endL);
                emit(elseL + ":");
                for (Stmt st : i.elseBlk()) genStmt(st);
                emit("  br label %" + endL);
                emit(endL + ":");
            } else if (s instanceof While w) {
                String condL = freshLabel("cond");
                String bodyL = freshLabel("body");
                String exitL = freshLabel("exit");
                emit("  br label %" + condL);
                emit(condL + ":");
                String c = genExpr(w.cond());
                String ci1 = fresh();
                emit("  " + ci1 + " = icmp ne i64 " + c + ", 0");
                emit("  br i1 " + ci1 + ", label %" + bodyL + ", label %" + exitL);
                emit(bodyL + ":");
                for (Stmt st : w.body()) genStmt(st);
                emit("  br label %" + condL);
                emit(exitL + ":");
            }
        }

        private String genExpr(Expr e) {
            if (e instanceof IntLit il) return Long.toString(il.value());
            if (e instanceof Var v) {
                String r = fresh();
                emit("  " + r + " = load i64, ptr %" + v.name());
                return r;
            }
            if (e instanceof BinOp b) {
                String lv = genExpr(b.left());
                String rv = genExpr(b.right());
                String res = fresh();
                Map<String, String> arith = Map.of(
                    "+", "add", "-", "sub", "*", "mul", "/", "sdiv", "%", "srem"
                );
                Map<String, String> cmp = Map.of(
                    "==", "eq", "!=", "ne", "<", "slt", "<=", "sle", ">", "sgt", ">=", "sge"
                );
                if (arith.containsKey(b.op())) {
                    emit("  " + res + " = " + arith.get(b.op()) + " i64 " + lv + ", " + rv);
                } else if (cmp.containsKey(b.op())) {
                    String tmp = fresh();
                    emit("  " + tmp + " = icmp " + cmp.get(b.op()) + " i64 " + lv + ", " + rv);
                    emit("  " + res + " = zext i1 " + tmp + " to i64");
                } else {
                    throw new RuntimeException("Unknown op: " + b.op());
                }
                return res;
            }
            throw new RuntimeException("Unknown expr: " + e);
        }
    }

    // ===== Main =====

    public static void main(String[] args) throws Exception {
        if (args.length == 0) {
            System.err.println("usage: java Main.java <source.nb>");
            System.exit(1);
        }
        String src = Files.readString(Path.of(args[0]));
        List<Token> tokens = tokenize(src);
        List<Stmt> program = new Parser(tokens).parseProgram();
        String ir = new CodeGen().gen(program);
        System.out.print(ir);
    }
}
